cpp/src/stream_compaction/distinct.cu

```diff
@@ -59,12 +60,25 @@ rmm::device_uvector<size_type> get_distinct_indices(table_view const& input,
   auto const row_hasher = cudf::experimental::row::hash::row_hasher(preprocessed_input);
   auto const key_hasher = experimental::compaction_hash(row_hasher.device_hasher(has_nulls));
 
-  auto const row_comp  = cudf::experimental::row::equality::self_comparator(preprocessed_input);
-  auto const key_equal = row_comp.equal_to(has_nulls, nulls_equal);
+  auto const row_comp = cudf::experimental::row::equality::self_comparator(preprocessed_input);
 
   auto const pair_iter = cudf::detail::make_counting_transform_iterator(
     size_type{0}, [] __device__(size_type const i) { return cuco::make_pair(i, i); });
-  map.insert(pair_iter, pair_iter + input.num_rows(), key_hasher, key_equal, stream.value());
+
+  auto const insert_keys = [&](auto const value_comp) {
+    auto const key_equal = row_comp.equal_to(has_nulls, nulls_equal, value_comp);
+    map.insert(pair_iter, pair_iter + input.num_rows(), key_hasher, key_equal, stream.value());
+  };
+
+  using nan_equal_comparator =
+    cudf::experimental::row::equality::nan_equal_physical_equality_comparator;
+  using nan_unequal_comparator = cudf::experimental::row::equality::physical_equality_comparator;
+
+  if (nans_equal == nan_equality::ALL_EQUAL) {
+    insert_keys(nan_equal_comparator{});
+  } else {
+    insert_keys(nan_unequal_comparator{});
+  }
```

Comment:
Define these comparator types in the corresponding branch of the `if` to increase the locality of the definition (defined just before use).
```suggestion
  if (nans_equal == nan_equality::ALL_EQUAL) {
    using nan_equal_comparator =
      cudf::experimental::row::equality::nan_equal_physical_equality_comparator;
    insert_keys(nan_equal_comparator{});
  } else {
    using nan_unequal_comparator = cudf::experimental::row::equality::physical_equality_comparator;
    insert_keys(nan_unequal_comparator{});
  }
```